### PR TITLE
reverse-proxy docs: add missing .well-known paths

### PR DESCRIPTION
**Pull request by wagner-intevation**
Add missing .well-known paths to Apache, nginx and HAProxy

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [ ] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: changelog.d/13820.doc [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Add missing .well-known paths to Apache, nginx and HAProxy
```
Comment (DMRobertson) [inline changelog.d/13820.doc:1]:
```suggestion
Add missing `.well-known` paths to Apache, nginx and HAProxy sample config. Contributed by Intevation GmbH.
```

**Review by DMRobertson — APPROVED** [review]
Thanks, this seems sane to me!

File: changelog.d/13820.doc (hunk 1 above)
Comment (DMRobertson) [reply changelog.d/13820.doc]: Argh, thanks. My mistake.

**Review by richvdh — CHANGES_REQUESTED** [review]
I don't think this is correct. Why is this being added?

In general it's better to serve the `.well-known` files from the reverse proxy rather than have synapse do it.

**Comment by richvdh** [comment]
see also https://github.com/matrix-org/synapse/pull/7077 which attempted to do something similar

**Comment by wagner-intevation** [comment]
https://matrix-org.github.io/synapse/latest/reverse_proxy.html contains configuration examples for different programs, but they are inconsistent. .well-known is missing in the Apache and nginx examples.

https://matrix-org.github.io/synapse/latest/setup/installation.html#client-well-known-uri contains a configuration example, but only for nginx and not for the other programs.

For Apache, returning strings form the configuration is not possible without using a [workaround using `ErrorDocument`](https://serverfault.com/a/1062092/217116).

Isn't it better to let the .well-known directory be served by synapse instead of returning a 404 by default? If that is not correct, would it be correct using the `ErrorDocument` workaround?

**Comment by richvdh** [comment]
> https://matrix-org.github.io/synapse/latest/reverse_proxy.html contains configuration examples for different programs, but they are inconsistent. .well-known is missing in the Apache and nginx examples.

They will still be inconsistent after your changes. The Caddy and haproxy examples (which I assume you're referring to) show how to configure delegation with those reverse proxies. They do *not* blindly forward `.well-known` to Synapse. Examples of how to do *the same thing* on Apache and nginx would be good to add.

> https://matrix-org.github.io/synapse/latest/setup/installation.html#client-well-known-uri contains a configuration example, but only for nginx and not for the other programs.

You don't appear to be suggesting we change this, though?

> For Apache, returning strings form the configuration is not possible without using a [workaround using `ErrorDocument`](https://serverfault.com/a/1062092/217116).

Why not just use an external file? Apache is pretty good at serving them up.

> Isn't it better to let the .well-known directory be served by synapse instead of returning a 404 by default? 

I don't think so, no. Synapse's support for `.well-known` files is pretty limited and only useful in a small number of cases. Indeed Synapse itself will serve a 404 for `.well-known/matrix/server` by default.

**Comment by wagner-intevation** [comment]
> You don't appear to be suggesting we change this, though?

Correct, In my message I made you aware of the difference, I did not make a suggestion.

> Why not just use an external file? Apache is pretty good at serving them up.

If that's the correct way to do it, I'll adapt the PR accordingly

**Comment by wagner-intevation** [comment]
The nginx section now contains the exact copy of what is in https://matrix-org.github.io/synapse/latest/setup/installation.html#client-well-known-uri for nginx and for apache2 I changed it to static files and a full example for the .well-known file as in https://matrix-org.github.io/synapse/latest/setup/installation.html#client-well-known-uri

File: docs/reverse_proxy.md [hunk 2]
```diff
@@ -80,6 +80,11 @@ server {
         # Increase client_max_body_size to match max_upload_size defined in homeserver.yaml
         client_max_body_size 50M;
     }
+    location /.well-known/matrix/client {
+        return 200 '{"m.homeserver": {"base_url": "https://<matrix.example.com>"}}';
```
Comment (richvdh) [inline docs/reverse_proxy.md:84]: we should probably make this match the example for Caddy, Apache and haproxy (ie, make it include `m.identity_server`)?

File: docs/reverse_proxy.md [hunk 3]
```diff
@@ -80,6 +80,11 @@ server {
...
+        default_type application/json;
+        add_header Access-Control-Allow-Origin *;
```
Comment (richvdh) [inline docs/reverse_proxy.md:86]: do we not need to set the other CORS headers here and for Apache? (https://spec.matrix.org/v1.4/client-server-api/#web-browser-clients)

**Review by richvdh — CHANGES_REQUESTED** [review]
If we're going to add this stuff here, we should explain what it is for and link to the full documentation (https://matrix-org.github.io/synapse/latest/setup/installation.html#client-well-known-uri), rather than just dumping some configuration settings and expecting people to cargo-cult it, and that it will work perfectly for them.

It looks like the Caddy and HAProxy examples link to the "delegation" documentation, but not the "client-well-known" documentation; that seems incorrect to me. Possibly they should say 

```
Examples for [server](delegate.md#well-known-delegation) and [client](setup/installation.html#client-well-known-url) `.well-known` files:
```

**Comment by clokep** [comment]
I'm going to close this pull request due to lack of response, please respond to the above feedback if you'd like to continue working on it. Thanks!

**Pull request closed without merging**